### PR TITLE
WIP: Normalize less during type-checking

### PR DESCRIPTION
**Pull request by sjakobi**
This is an experiment to find out whether we can get away with fewer normalization steps during typechecking.

---
TODO:

* [ ] Add a property test to check that we infer the same types for expressions containing `Note`s as for their de-noted counterparts.
* [ ] Test whether an expressions' type is now generally equal to the type inferred from the normalized expression.
* [ ] Fix type inference for annotated `merge`s.

### CODE REVIEW

**Comment by sjakobi** [comment]
The Haskell implementation already passes the new test case.

**Comment by sjakobi** [comment]
I should add a property test that the output of typeWithA is always normalized.

File: dhall/src/Dhall/TypeCheck.hs [hunk 1]
```diff
@@ -117,26 +117,26 @@ typeWithA tpa = loop
                 return a
     loop ctx   (Lam x _A  b     ) = do
         _ <- loop ctx _A
-        let ctx' = fmap (Dhall.Core.shift 1 (V x 0)) (Dhall.Context.insert x (Dhall.Core.normalize _A) ctx)
+        let ctx' = fmap (Dhall.Core.shift 1 (V x 0)) (Dhall.Context.insert x (Dhall.Core.normalize _A) ctx) -- TODO: Add test case
         _B <- loop ctx' b
-        let p = Pi x _A _B
+        let p = Pi x (Dhall.Core.normalize _A) _B -- TODO: Add test
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]:
```
    Inferred types should be normalized: FAIL (0.28s)
      *** Failed! Falsifiable (after 54636 tests and 2 shrinks):
      Lam "" (Let "" Nothing NaturalEven Text) (Const Type)
      Got: Right (Pi "" (Let "" Nothing NaturalEven Text) (Const Kind))
      Expected: Right (Pi "" Text (Const Kind))
      Use --quickcheck-replay=737763 to reproduce.
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 2]
```diff
@@ -463,12 +464,12 @@ typeWithA tpa = loop
...
         let process k v = do
                 t <- loop ctx v
-                s <- fmap Dhall.Core.normalize (loop ctx t)
+                s <- loop ctx t
                 case s of
                     Const _ -> return t
                     _ -> Left (TypeError ctx e (InvalidFieldType k t))
 
-        Record <$> Dhall.Map.unorderedTraverseWithKey process kvs
+        Record <$> Dhall.Map.unorderedTraverseWithKey process (Dhall.Map.sort kvs) -- TODO: Add test
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]:
```
    Inferred types should be normalized: FAIL (0.01s)
      *** Failed! Falsifiable (after 1268 tests and 10 shrinks):
      Combine (RecordLit (fromList [("a",IntegerLit 0)])) (RecordLit (fromList [("",BoolLit False)]))
      Use --quickcheck-replay=207379 to reproduce.
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 3]
```diff
@@ -117,33 +117,33 @@ typeWithA tpa = loop
...
-        tf <- fmap Dhall.Core.normalize (loop ctx f)
+        tf <- fmap Dhall.Core.normalize (loop ctx f) -- checked via customization/doubleReduction
         (x, _A, _B) <- case tf of
             Pi x _A _B -> return (x, _A, _B)
             _          -> Left (TypeError ctx e (NotAFunction f tf))
         _A' <- loop ctx a
         if Dhall.Core.judgmentallyEqual _A _A'
             then do
-                let a'   = Dhall.Core.shift   1  (V x 0) a
+                let a'   = Dhall.Core.shift   1  (V x 0) (Dhall.Core.normalize a) -- TODO: Add test
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]:
```
    Inferred types should be normalized: FAIL (6.63s)
      *** Failed! Falsifiable (after 1420653 tests and 2 shrinks):
      App ListHead (Let "" Nothing (IntegerLit 0) Integer)
      Got: Right (Pi "_" (App List (Let "" Nothing (IntegerLit 0) Integer)) (App Optional (Let "" Nothing (IntegerLit 0) Integer)))
      Expected: Right (Pi "_" (App List Integer) (App Optional Integer))
      Use --quickcheck-replay=926229 to reproduce.
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 4]
```diff
@@ -178,7 +179,7 @@ typeWithA tpa = loop
         t' <- loop ctx x
         if Dhall.Core.judgmentallyEqual t t'
             then do
-                return t
+                return t' -- TODO: Update standard
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]:
```
    Inferred types should be normalized: FAIL (6.63s)
      *** Failed! Falsifiable (after 1420653 tests and 2 shrinks):
      App ListHead (Let "" Nothing (IntegerLit 0) Integer)
      Got: Right (Pi "_" (App List (Let "" Nothing (IntegerLit 0) Integer)) (App Optional (Let "" Nothing (IntegerLit 0) Integer)))
      Expected: Right (Pi "_" (App List Integer) (App Optional Integer))
      Use --quickcheck-replay=926229 to reproduce.
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 5]
```diff
@@ -117,33 +117,34 @@ typeWithA tpa = loop
                 return a
     loop ctx   (Lam x _A  b     ) = do
         _ <- loop ctx _A
-        let ctx' = fmap (Dhall.Core.shift 1 (V x 0)) (Dhall.Context.insert x (Dhall.Core.normalize _A) ctx)
+        let nf_A = Dhall.Core.normalize _A
+        let ctx' = fmap (Dhall.Core.shift 1 (V x 0)) (Dhall.Context.insert x nf_A ctx) -- TODO: Add test case
         _B <- loop ctx' b
-        let p = Pi x _A _B
+        let p = Pi x nf_A _B -- TODO: Add test
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]: Inferred types should be normalized: FAIL (0.28s)
      *** Failed! Falsifiable (after 54636 tests and 2 shrinks):
      Lam "" (Let "" Nothing NaturalEven Text) (Const Type)
      Got: Right (Pi "" (Let "" Nothing NaturalEven Text) (Const Kind))
      Expected: Right (Pi "" Text (Const Kind))
      Use --quickcheck-replay=737763 to reproduce.

File: dhall/tests/Dhall/Test/QuickCheck.hs [hunk 6]
```diff
@@ -397,6 +410,10 @@ tests =
           , Test.QuickCheck.property isSameAsSelf
           , QuickCheckTests 10000
           )
+        , ( "Inferred types should be normalized"
```
Comment (Gabriel439) [inline dhall/tests/Dhall/Test/QuickCheck.hs]: Note that this is currently not one of the four soundness properties that the language aims to enforce:

https://github.com/dhall-lang/dhall-lang/blob/5fea4544f6fe57acc034828a0a6ccebfd52aad15/.github/CONTRIBUTING.md#soundness-of-proposed-changes

Specifically, the language currently only guarantees that inferred types are safe to normalize (i.e. the inferred type does not itself need to type-checked again before being normalized).  We could add the guarantee that you describe, but that guarantee should be upstreamed into the standard.

Comment (Gabriel439) [reply dhall/tests/Dhall/Test/QuickCheck.hs]: Also, such a guarantee would imply other changes to the standard (i.e. we would have to replace all occurrences of `:→` with `:`, for example)

Comment (sjakobi) [reply dhall/tests/Dhall/Test/QuickCheck.hs]: Thanks for the pointer to the soundness properties! I have opened https://github.com/dhall-lang/dhall-lang/issues/724 to discuss the potential standard change.

**Comment by sjakobi** [comment]
I can also remove the `normalize` step in the handler for the `Type` command.

File: dhall/src/Dhall/TypeCheck.hs [hunk 7]
```diff
@@ -118,33 +118,34 @@ typeWithA tpa = loop
...
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]: This would require an additional normalization in the standard.

File: dhall/src/Dhall/TypeCheck.hs [hunk 8]
```diff
@@ -675,7 +677,7 @@ typeWithA tpa = loop
         Data.Foldable.traverse_ (loop ctx) mT₁
 
         let ktX = appEndo (foldMap (Endo . compareFieldTypes) ktsX) Nothing
-            mT₂ = fmap Dhall.Core.normalize mT₁
+            mT₂ = fmap Dhall.Core.normalize mT₁ -- TODO: Add test
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]: https://github.com/dhall-lang/dhall-lang/pull/744

**Comment by sjakobi** [comment]
Inferring the type for an example dhall-kubernetes deployment takes about 30% less time now:

With `master`:

```
≻ bench "dhall type --file examples/deployment.dhall > /dev/null"
benchmarking dhall type --file examples/deployment.dhall > /dev/null
time                 1.564 s    (1.561 s .. 1.566 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 1.565 s    (1.564 s .. 1.566 s)
std dev              978.9 μs   (0.0 s .. 1.065 ms)
variance introduced by outliers: 19% (moderately inflated)
```

With this branch:

```
≻ bench "dhall type --file examples/deployment.dhall > /dev/null"
benchmarking dhall type --file examples/deployment.dhall > /dev/null
time                 1.032 s    (1.013 s .. 1.054 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 1.027 s    (1.023 s .. 1.029 s)
std dev              3.908 ms   (0.0 s .. 4.469 ms)
variance introduced by outliers: 19% (moderately inflated)
```

**Comment by sjakobi** [comment]
There's still a bug here:

```
$ echo 'merge { x = 0 } < x >.x : Bool' | cabal run -- dhall:dhall type
Natural
```

But for some reason the `MergeAnnotationMismatch` failure test doesn't complain. It's not the caches.

**Comment by sjakobi** [comment]
> There's still a bug here:

Fixed in https://github.com/dhall-lang/dhall-haskell/pull/1326.

**Comment by sjakobi** [comment]
Superseded by #1337. I'll submit separate PRs for some of the smaller fixes.

**Pull request closed without merging**